File: analise-busca-binaria.py
import random
import time

def gerar_sequencia(n):
    return [random.randint(0, 50) for _ in range(n)]

def gerar_chave(q):
    return [random.randint(0, 50) for _ in range(q)]

# Métodos de busca para os experimentos
def busca_sequencial(V, chave):
    for i, elemento in enumerate(V):
        if elemento == chave:
            return i
    return -1

def busca_sequencial_otimizada(V, chave):
    for i, elemento in enumerate(V):
        if elemento > chave:
            return -1
        if elemento == chave:
            return i
    return -1

def busca_binaria(V, chave):
    inicio = 0
    fim = len(V) - 1
    while inicio <= fim:
        meio = (inicio + fim) // 2
        if V[meio] == chave:
            return meio
        elif V[meio] < chave:
            inicio = meio + 1
        else:
            fim = meio - 1
    return -1
    
# Função para executar os experimentos
def executar_experimentos(n, q, num_exec):
    sequencia = gerar_sequencia(n)
    chaves = gerar_chave(q)

    sequencia.sort()

    def medir_tempo(busca):
        tempo_inicio = time.time()
        for chave in chaves:
            busca(sequencia, chave)
        return (time.time() - tempo_inicio) / num_exec

    tempo_busca_sequencial = medir_tempo(busca_sequencial)
    tempo_busca_sequencial_otimizada = medir_tempo(busca_sequencial_otimizada)
    tempo_busca_binaria = medir_tempo(busca_binaria)

    print("# n =", n)
    print("# q =", q)
    print("# num_exec =", num_exec)
    print("# busca_sequencial =", tempo_busca_sequencial)
    print("# busca_sequencial_otimizada =", tempo_busca_sequencial_otimizada)
    print("# busca_binaria =", tempo_busca_binaria)
    print("\n")

# Parâmetros do experimento
valores_n = [10**4, 10**5, 10**6, 10**7]
valores_q = [10**2, 10**3, 10**4, 10**5]

# Número de execuções para calcular o tempo médio
num_exec = 5 

# Execução do experimento
for n in valores_n:
    for q in valores_q:
        executar_experimentos(n, q, num_exec)

File: analise-busca-binaria.cpp
#include <iostream>
#include <vector>
#include <cstdlib>
#include <ctime>
#include <algorithm>

// Função de busca sequencial
int busca_sequencial(std::vector<int>& V, int x) {
    for (int i = 0; i < V.size(); i++) {
        if (V[i] == x) {
            return i;
        }
    }
    return -1;
}

// Função de busca sequencial otimizada
int busca_sequencial_otimizada(std::vector<int>& V, int x) {
    for (int i = 0; i < V.size(); i++) {
        if (V[i] == x) {
            return i;
        }
        else if (V[i] > x) {
            return -1;
        }
    }
    return -1;
}

// Função de busca binária
int busca_binaria(std::vector<int>& V, int x) {
    int baixo = 0, alto = V.size() - 1;
    while (baixo <= alto) {
        int meio = (baixo + alto) / 2;
        if (V[meio] == x) {
            return meio;
        }
        else if (V[meio] < x) {
            baixo = meio + 1;
        }
        else {
            alto = meio - 1;
        }
    }
    return -1;
}

// Função para gerar uma sequência aleatória
void gerar_sequencia(std::vector<int>& V, int n) {
    V.clear(); // Limpa o vetor, se houver algum conteúdo anterior
    for (int i = 0; i < n; i++) {
        V.push_back(rand() % 101); // números aleatórios entre 0 e 100
    }
}

// Gerar chaves aleatórias
void gerar_chave(std::vector<int>& V, int q) {
    V.clear(); // Limpa o vetor, se houver algum conteúdo anterior
    for (int i = 0; i < q; i++) {
        V.push_back(rand() % 101); // números aleatórios entre 0 e 100
    }
}

// Função para executar uma busca e medir o tempo
double executar_busca(std::vector<int>& V, int q, int tipo_busca) {
    clock_t inicio, fim;
    double tempo_total = 0;

    // Realizar 1000 execuções e calcular o tempo médio
    for (int k = 0; k < 1000; k++) {
        // Gerar chaves aleatórias
        gerar_chave(V, q);

        inicio = clock();
        for (int i = 0; i < q; i++) {
            if (tipo_busca == 0) {
                busca_sequencial(V, V[i]);
            }
            else if (tipo_busca == 1) {
                busca_sequencial_otimizada(V, V[i]);
            }
            else {
                busca_binaria(V, V[i]);
            }
        }
        fim = clock();

        tempo_total += ((double) (fim - inicio)) / CLOCKS_PER_SEC;
    }

    return tempo_total / 1000;
}

int main() {
    std::vector<int> sequencia;
    std::vector<int> chaves;

    // Tamanho da sequência e número de consultas
    std::vector<int> tamanhos_sequencia = {10000, 100000, 1000000, 10000000};
    std::vector<int> num_consultas = {100, 1000, 10000, 100000};

    for (int i = 0; i < tamanhos_sequencia.size(); i++) {
        for (int j = 0; j < num_consultas.size(); j++) {
            int n = tamanhos_sequencia[i];
            int q = num_consultas[j];

            // Gerar sequência aleatória
            gerar_sequencia(sequencia, n);

            // Busca sequencial
            double tempo_medio_seq = executar_busca(sequencia, q, 0);

            // Busca sequencial otimizada
            double tempo_medio_seq_otim = executar_busca(sequencia, q, 1);

            // Ordenar sequência para busca binária
            std::sort(sequencia.begin(), sequencia.end());

            // Busca binária
            double tempo_medio_binaria = executar_busca(sequencia, q, 2);

            std::cout << "Experimentos para n = " << n << ", q = " << q << "\n";
            std::cout << "Tempo médio da busca sequencial: " << tempo_medio_seq << " segundos\n";
            std::cout << "Tempo médio da busca sequencial otimizada: " << tempo_medio_seq_otim << " segundos\n";
            std::cout << "Tempo médio da busca binária: " << tempo_medio_binaria << " segundos\n";
        }
    }

    return 0;
}
